compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java

```diff
@@ -220,15 +221,33 @@ public static BLangClassDefinition createClassDef(Location pos, BObjectTypeSymbo
         return classDefNode;
     }
 
-    public static BLangErrorType createBLangErrorType(Location pos, BType detailType) {
+    public static BLangErrorType createBLangErrorType(Location pos, BType detailType, SymbolEnv env, Names names) {
         BLangErrorType errorType = (BLangErrorType) TreeBuilder.createErrorTypeNode();
         BLangUserDefinedType userDefinedTypeNode = (BLangUserDefinedType) TreeBuilder.createUserDefinedTypeNode();
         userDefinedTypeNode.pos = pos;
         userDefinedTypeNode.pkgAlias = (BLangIdentifier) TreeBuilder.createIdentifierNode();
-        userDefinedTypeNode.typeName = createIdentifier(pos, detailType.tsymbol.name.value);
+
+        // When the error detail type is a inline type decl, such as in `error<map<value:Cloneable>>` we need to
+        // add a explicit type decl for the detail type.
```

Comment:
No, actually we need the anon name when the detail type is a map, records already has a user-defined or anon type-def.

Reply:
Ah sorry, my bad, misread the flow from `defineIntersectionType`. 

But for maps, shouldn't we use a map type (`BLangConstrainedType`?) as the detail type instead of creating a user-defined type? IINM, that's what we do for user-defined error type definitions with maps too?

Reply:
Let me check that

Reply:
It turned out that we still have to create user defined type for T where T is the type constraint.

Reply:
Is this specifically for intersection types? For intersection types I thought we usually create a record. Can you please share an example where we get to this point for a map that would then require a user-defined type? I tried several combinations with maps as details types but they seem to work for me.

For something like `type MyError error<map<anydata>>;` we don't seem to be creating a user-defined type?

Reply:
Can you try intersecting errors created from build-in error like in this issue https://github.com/ballerina-platform/ballerina-lang/issues/28563
At the moment my code is bit different from the PR will provide you a sample as soon as I push my changes

Reply:
Thanks!

So, with the original issue, the NPE seems to be because the initializer function is null at https://github.com/ballerina-platform/ballerina-lang/blob/v2.0.0-alpha5/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java#L1898, right? I'm not sure if depending on `org.wso2.ballerinalang.compiler.semantics.model.symbols.BStructureTypeSymbol#initializerFunc` of a record symbol here is correct actually, don't we generally create it at desugar? Wonder if we can avoid using it here. Or if it has to be present with intersection types, maybe we've missed creating it somewhere?

Tried with 
```ballerina
public type Foo distinct error;

public type BarData record {
    error?[] errs = [];
};

public type Bar distinct Foo & error<BarData>;
```

But with an intersection with the builtin error like
```ballerina
public type Foo distinct error<BarData>;

public type BarData record {
    error?[] errs = [];
};

public type Bar error & Foo;
```
when we get to https://github.com/ballerina-platform/ballerina-lang/blob/v2.0.0-alpha5/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java#L228 it seems like `detailType` is `noType`, which is wrong, right? Could that be the issue here rather than it being a map?